migrations/2023_07_05_000000_alter_polls_use_post_id.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+/*
+ * This file is part of fof/polls.
+ *
+ * Copyright (c) FriendsOfFlarum.
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+use Illuminate\Database\Query\JoinClause;
+use Illuminate\Database\Schema\Blueprint;
+use Illuminate\Database\Schema\Builder;
+
+return [
+    'up' => function (Builder $schema) {
+        $db = $schema->getConnection();
+
+        $schema->table('polls', function (Blueprint $table) {
+            $table->dropForeign(['discussion_id']);
+        });
+
+        $db->transaction(function () use ($db) {
+            // Don't run through this step if no rows exist in the polls table
+            if (!$db->table('polls')->exists()) {
+                return;
+            }
+
+            // Update polls whose discussions have a clear first post ID associated
+            $db->table('polls')
+                ->join('discussions', function (JoinClause $join) {
+                    $join->on('polls.discussion_id', '=', 'discussions.id')
+                        ->where('discussions.first_post_id', '!=', null);
+                })
+                ->update(['polls.discussion_id' => $db->raw('discussions.first_post_id')]);
+
+            // Update polls whose discussions have a null first post ID associated
+            $firstPosts = $db->table('posts')
+                ->where('number', '=', 1);
+
+            $db->table('polls')
+                ->join('discussions', function (JoinClause $join) {
+                    $join->on('polls.discussion_id', '=', 'discussions.id')
+                        ->where('discussions.first_post_id', '=', null);
+                })
+                ->leftJoinSub($firstPosts, 'first_posts', function (JoinClause $join) {
+                    $join->on('first_posts.discussion_id', '=', 'discussions.id');
+                })
+                ->update(['polls.discussion_id' => $db->raw('first_posts.id')]);
+
+            // Delete polls that don't have an associated post
+            // TODO Is this a good idea? Not sure what situations can result in first_post_id being null but the discussion still existing with an associated first post.
```

Comment:
The previous two updates should take care of
- existing first post relation through `first_post_id`
- post with `number = 1` 

I'm very much not an SQL expert, and I have barely touched joins myself. I don't trust myself to merge this without a lot of testing done to avoid polls disappearing that shouldn't.

Reply:
From further testing, the second scenario (no first post relation but post with number ID 1) cannot be achieved simply by deleting the first post in the discussion. If there are other posts, they retain their numbering (i.e. post numbering starts at 2). Extensions such as `fof/merge-discussions` and others renumber posts, in which case the poll would reappear (in v1).

I don't have a problem with keeping that scenario. Just providing some more info, since a poll is currently only shown for post number #1.